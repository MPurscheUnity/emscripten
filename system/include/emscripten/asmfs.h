/*
 * Copyright 2020 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

#include <stdint.h>
#include <stddef.h>

#ifdef __cplusplus
extern "C" {
#endif

#define EMSCRIPTEN_ASMFS_FILE_RESIDENCY_FLAGS unsigned int
#define EMSCRIPTEN_ASMFS_WASM_RESIDENT 1
#define EMSCRIPTEN_ASMFS_JS_RESIDENT 2
#define EMSCRIPTEN_ASMFS_INDEXEDDB_RESIDENT 4
#define EMSCRIPTEN_ASMFS_LOCALSTORAGE_RESIDENT 8
#define EMSCRIPTEN_ASMFS_SESSIONSTORAGE_RESIDENT 16

typedef void *AsmfsFileHandle;

typedef void (*asmfs_fetch_finished_callback)(int httpStatusResult, AsmfsFileHandle fileHandle, void *userData);

// Starts an asynchronous download of the given URL, and places the resulting ArrayBuffer to the given destination location in the filesystem, backed by JS side memory.
extern void emscripten_asmfs_async_file_fetch(const char *url, const char *dst, EMSCRIPTEN_ASMFS_FILE_RESIDENCY_FLAGS residencyFlags, asmfs_fetch_finished_callback cb, void *userData);

extern uint8_t *emscripten_asmfs_get_file_ptr(AsmfsFileHandle fileHandle);

extern size_t emscripten_asmfs_get_file_size(AsmfsFileHandle fileHandle);

extern const char *emscripten_asmfs_get_file_name(AsmfsFileHandle fileHandle);

extern void emscripten_asmfs_dump_tree(void);

extern size_t emscripten_asmfs_copy_file_chunk_to_wasm_memory(AsmfsFileHandle fileHandle, size_t srcOffset, void *dst, size_t dstLength);

// Brings the file from JavaScript side residency over to WebAssembly heap.
// Note: If a copy of the file already exists in the wasm heap, it is freed, and
// the file contents are re-copied(!) to the Wasm heap again from JavaScript side
// backing. This is to support the use case where the file is brought to Wasm side,
// modified, but the modifications are to be discarded/replaced from the JS side
// copy. 
extern uint8_t *emscripten_asmfs_copy_file_to_wasm_memory(AsmfsFileHandle fileHandle);

extern void emscripten_asmfs_copy_file_to_js_memory(AsmfsFileHandle fileHandle);

extern void emscripten_asmfs_discard_file_from_wasm_memory(AsmfsFileHandle fileHandle);

extern EMSCRIPTEN_ASMFS_FILE_RESIDENCY_FLAGS emscripten_asmfs_get_file_residency(AsmfsFileHandle fileHandle);

#ifdef __cplusplus
}
#endif
