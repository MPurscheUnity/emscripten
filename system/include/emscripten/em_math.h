/*
 * Copyright 2020 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

#ifdef __cplusplus
extern "C" {
#endif

/**
 * This file contains C functions to access the JavaScript Math API via Emscripten.
 * Please note that accessing these functions is relatively slow, since they each
 * incur a language boundary crossing call from WebAssembly out to JavaScript.

 * These functions are best used in scenarios where small code size is more desirable
 * than performance.

 * See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math
 * for details. */

// The following operations have very fast WebAssembly opcodes. Therefore they are not
// exposed as individual functions:

// Math.abs(x) -> f32.abs and f64.abs. (use fabsf() and fabs() from math.h)
// Math.ceil -> f32.ceil and f64.ceil (ceil() and ceilf() in math.h)
// Math.clz32(x) -> i32.clz and i64.clz (call __builtin_clz() and __builtin_clzll())
// Math.floor -> f32.floor and f64.floor (floor() and floorf() in math.h)
// Math.fround -> f64.promote_f32(f32.demote_f64()) (call double d = (double)(float)someDouble;)
// Math.imul(x, y) -> i32.mul and i64.mul (directly multiply two signed integers)
// Math.min -> f32.min and f64.min (fminf() and fmin() in math.h)
// Math.max -> f32.max and f64.max (fmaxf() and fmax() in math.h)
// Math.trunc -> f32.trunc and f64.trunc (truncf() and trunc() in math.h)

// The following constants are available on the JS Math object, mirrored here for convenience.

#define EM_MATH_E 2.718281828459045
#define EM_MATH_LN2 0.6931471805599453
#define EM_MATH_LN10 2.302585092994046
#define EM_MATH_LOG2E 1.4426950408889634
#define EM_MATH_LOG10E 0.4342944819032518
#define EM_MATH_PI 3.141592653589793
#define EM_MATH_SQRT1_2 0.7071067811865476
#define EM_MATH_SQRT2 1.4142135623730951

// The following Math operations do not have native WebAssembly opcodes, and
// are provided here as small sized alternatives to their libc counterparts.

double emscripten_math_acos(double x) __attribute__((const)); // acos() in math.h
double emscripten_math_acosh(double x) __attribute__((const)); // acosh() in math.h
double emscripten_math_asin(double x) __attribute__((const)); // asin() in math.h
double emscripten_math_asinh(double x) __attribute__((const)); // asinh() in math.h
double emscripten_math_atan(double x) __attribute__((const)); // atan() in math.h
double emscripten_math_atan2(double y, double x) __attribute__((const)); // atan2() in math.h
double emscripten_math_atanh(double x) __attribute__((const)); // atanh() in math.h
double emscripten_math_cbrt(double x) __attribute__((const)); // cbrt() in math.h
double emscripten_math_cos(double x) __attribute__((const)); // cos() in math.h
double emscripten_math_cosh(double x) __attribute__((const)); // cosh() in math.h
double emscripten_math_exp(double x) __attribute__((const)); // exp() in math.h
double emscripten_math_expm1(double x) __attribute__((const)); // expm1() in math.h
double emscripten_math_fmod(double x, double y) __attribute__((const)); // fmod() in math.h, not a function on Math, but calls JS "x % y" operator.
double emscripten_math_hypot(int count, ...) __attribute__((const));  // hypot() in math.h (although only for fixed 2 arguments)
double emscripten_math_log(double x) __attribute__((const)); // log() in math.h
double emscripten_math_log1p(double x) __attribute__((const)); // log1p() in math.h
double emscripten_math_log10(double x) __attribute__((const)); // log10() in math.h
double emscripten_math_log2(double x) __attribute__((const)); // log2() in math.h
double emscripten_math_pow(double x, double y) __attribute__((const)); // pow(x, y) in math.h
double emscripten_math_random(void); // N.b. The difference between this function and emscripten_random() in emscripten.h is that it returns a single-precision float!
double emscripten_math_round(double x) __attribute__((const)); // round() in math.h
double emscripten_math_sign(double x) __attribute__((const)); // No equivalent in libc
double emscripten_math_sin(double x) __attribute__((const)); // sin() in math.h
double emscripten_math_sinh(double x) __attribute__((const)); // sinh() in math.h
double emscripten_math_sqrt(double x) __attribute__((const)); // sqrt() in math.h
double emscripten_math_tan(double x) __attribute__((const)); // tan() in math.h
double emscripten_math_tanh(double x) __attribute__((const)); // tanh() in math.h

#ifdef __cplusplus
}
#endif
