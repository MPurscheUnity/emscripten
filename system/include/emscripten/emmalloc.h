#pragma once

#ifdef __cplusplus
extern "C" {
#endif

// emmalloc: A lightweight web-friendly memory allocator suitable for very small applications.
// Enable the usage of emmalloc by passing the linker flag -s MALLOC=emmalloc to the application.

// A debug function that dumps the whole structure of malloc internal memory blocks to console.
// *extremely slow*, use for debugging allocation test cases.
void emmalloc_dump_memory_regions(void);

// Allocates size bytes with the given pow-2 alignment.
void *memalign(size_t alignment, size_t size);
void *emmalloc_memalign(size_t alignment, size_t size);
void *emscripten_builtin_memalign(size_t alignment, size_t size);
void *aligned_alloc(size_t alignment, size_t size);

// Allocates size bytes with default alignment (8 bytes)
void *malloc(size_t size);
void *emmalloc_malloc(size_t size);
void *emscripten_builtin_malloc(size_t size);

// Frees a memory pointer allocated with any of
// emmalloc_memalign, emmalloc_malloc, 
void free(void *ptr);
void emmalloc_free(void *ptr);
void emscripten_builtin_free(void *ptr);

// Performs a reallocation of the given memory pointer to a new size. If the memory region
// pointed by ptr cannot be resized in place, a new memory region will be allocated, old
// memory copied over, and the old memory area freed. The pointer ptr must have been
// allocated with one of the emmalloc memory allocation functions (malloc, memalign, ...).
// If called with size == 0, the pointer ptr is freed, and a null pointer is returned. If
// called with null ptr, a new pointer is allocated.
void *realloc(void *ptr, size_t size);
void *emmalloc_realloc(void *ptr, size_t size);

// emmalloc_realloc_try() is like realloc(), but only attempts to try to resize the existing
// memory area. If resizing the existing memory area fails, then realloc_try() will return 0
// (the original memory block is not freed or modified). If resizing succeeds, previous
// memory contents will be valid up to min(old length, new length) bytes.
// If a null pointer is passed, no allocation is attempted but the function will return 0.
// If zero size is passed, the function will behave like free().
void *emmalloc_realloc_try(void *ptr, size_t size);

// emmalloc_realloc_uninitialized() is like realloc(), but old memory contents
// will be undefined after reallocation. (old memory is not preserved in any case)
void *emmalloc_realloc_uninitialized(void *ptr, size_t size);

// Like realloc(), but allows specifying the alignment to allocate to. This function cannot
// be used to change the alignment of an existing allocation, but the original pointer should
// be aligned to the given alignment already. 
void *aligned_realloc(void *ptr, size_t alignment, size_t size);
void *emmalloc_aligned_realloc(void *ptr, size_t alignment, size_t size);

// emmalloc_aligned_realloc_uninitialized() is like aligned_realloc(), but old memory contents
// will be undefined after reallocation. (old memory is not preserved in any case)
void *emmalloc_aligned_realloc_uninitialized(void *ptr, size_t alignment, size_t size);

// posix_memalign allocates memory with a given alignment, like memalign, but with a slightly
// different usage signature.
int posix_memalign(void **memptr, size_t alignment, size_t size);
int emmalloc_posix_memalign(void **memptr, size_t alignment, size_t size);

// calloc allocates memory that is initialized to zero.
void *calloc(size_t num, size_t size);
void *emmalloc_calloc(size_t num, size_t size);

// mallinfo() returns information about current emmalloc allocation state. This function
// is very slow, only good for debugging. Avoid calling it for "routine" diagnostics.
struct mallinfo mallinfo();
struct mallinfo emmalloc_mallinfo();

// malloc_trim() frees up unused memory back to the system. Note: currently not implemented,
// but only provided for cross-compilation compatibility. Always returns 0.
int malloc_trim(size_t pad);
int emmalloc_malloc_trim(size_t pad);

#ifdef __cplusplus
}
#endif
