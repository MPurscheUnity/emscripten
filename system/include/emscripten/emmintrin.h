#pragma once

#include <xmmintrin.h>

typedef int32x4 __m128i;
typedef float64x2 __m128d;

static __inline__ __m128i __attribute__((__always_inline__))
_mm_set_epi32(int z, int y, int x, int w)
{
  return (__m128i){ w, x, y, z };
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_set1_epi32(int w)
{
  return (__m128i){ w, w, w, w };
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_setzero_si128()
{
  return (__m128i){ 0, 0, 0, 0 };
}

static __inline__ void __attribute__((__always_inline__))
_mm_store_si128(__m128i *p, __m128i a)
{
  *p = a;
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_and_si128(__m128i a, __m128i b)
{
  return a & b;
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_andnot_si128(__m128i a, __m128i b)
{
  return ~a & b;
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_or_si128(__m128i a, __m128i b)
{
  return a | b;
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_xor_si128(__m128i a, __m128i b)
{
  return a ^ b;
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_add_epi32(__m128i a, __m128i b)
{
  return a + b;
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_sub_epi32(__m128i a, __m128i b)
{
  return a - b;
}

#define _mm_shuffle_epi32(a, imm) __extension__ ({ \
  __m128i __a = (a); \
  (__m128i)__builtin_shufflevector((__v4si)__a, (__v4si) _mm_set1_epi32(0), \
                                   (imm) & 0x3, ((imm) & 0xc) >> 2, \
                                   ((imm) & 0x30) >> 4, ((imm) & 0xc0) >> 6); })

static __inline__ __m128 __attribute__((__always_inline__))
_mm_castsi128_ps(__m128i a)
{
  return emscripten_float32x4_fromInt32x4Bits(a);
}

static __inline__ __m128 __attribute__((__always_inline__))
_mm_cvtepi32_ps(__m128i a)
{
  return emscripten_float32x4_fromInt32x4(a);
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_castps_si128(__m128 a)
{
  return emscripten_int32x4_fromFloat32x4Bits(a);
}

static __inline__ __m128i __attribute__((__always_inline__))
_mm_cvtps_epi32(__m128 a)
{
  return emscripten_int32x4_fromFloat32x4(a);
}

static __inline__ __m128 __attribute__((__always_inline__))
_mm_move_sd(__m128d __a, __m128d __b)
{
  return __builtin_shufflevector(__a, __b, 2, 1);
}
