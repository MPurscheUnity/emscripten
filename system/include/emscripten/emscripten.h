/**
 * This file contains a few useful things for compiling C/C++ code
 * with Emscripten, an LLVM-to-JavaScript compiler.
 *
 * The code can be used permissively under the MIT license.
 *
 * http://emscripten.org
 */

#ifdef __cplusplus
extern "C" {
#endif

/*
 * Forces LLVM to not dead-code-eliminate a function. Note that
 * closure may still eliminate it at the JS level, for which you
 * should use EXPORTED_FUNCTIONS (see settings.js).
 *
 * Example usage:
 *   void EMSCRIPTEN_KEEPALIVE my_function() { .. }
 */
#define EMSCRIPTEN_KEEPALIVE __attribute__((used))

/*
 * Interface to the underlying JS engine. This function will
 * eval() the given script.
 */
extern void emscripten_run_script(const char *script);
extern int emscripten_run_script_int(const char *script);
extern void emscripten_async_run_script(const char *script, int millis);

/*
 * Set a C function as the main event loop. The JS environment
 * will call that function at a specified number of frames per
 * second. Setting 0 or a negative value as the fps will use
 * the browser's requestAnimationFrame mechanism.
 *
 * Pausing and resuming the main loop is useful if your app
 * needs to perform some synchronous operation, for example
 * to load a file from the network. It might be wrong to
 * run the main loop before that finishes (the original
 * code assumes that), so you can break the code up into
 * asynchronous callbacks, but you must pause the main
 * loop until they complete.
 */
extern void emscripten_set_main_loop(void (*func)(), int fps);
extern void emscripten_pause_main_loop();
extern void emscripten_resume_main_loop();
extern void emscripten_cancel_main_loop();

/*
 * Call a C function asynchronously, that is, after returning
 * control to the JS event loop. This is done by a setTimeout.
 * When building natively this becomes a simple direct call,
 * after SDL_Delay (you must include SDL.h for that).
 *
 * If millis is negative, the browser's requestAnimationFrame
 * mechanism is used.
 */
#if EMSCRIPTEN
extern void emscripten_async_call(void (*func)(), int millis);
#else
inline void emscripten_async_call(void (*func)(), int millis) {
  SDL_Delay(millis);
  func();
}
#endif

/*
 * Hide the OS mouse cursor over the canvas. Note that SDL's
 * SDL_ShowCursor command shows and hides the SDL cursor, not
 * the OS one. This command is useful to hide the OS cursor
 * if your app draws its own cursor.
 */
void emscripten_hide_mouse();

/*
 * Resizes the pixel width and height of the <canvas> element
 * on the Emscripten web page.
 */
void emscripten_set_canvas_size(int width, int height);

/*
 * Returns the highest-precision representation of the
 * current time that the browser provides. This uses either
 * Date.now or performance.now. The result is *not* an
 * absolute time, and is only meaningful in comparison to
 * other calls to this function. The unit is ms.
 */
float emscripten_get_now();

/*
 * Simple random number generation in [0, 1), maps to Math.random().
 */
float emscripten_random();

/*
 * Registers a callback function for receiving browser-generated keyboard input events.
 * See https://developer.mozilla.org/en/DOM/Event/UIEvent/KeyEvent
 * and http://www.javascriptkit.com/jsref/eventkeyboardmouse.shtml
 *
 * The contents of the event objects are passed to a C callback function of form
 *    void OnKeyboardEvent(int eventType, int charCode, int keyCode, int which, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, bool repeat);
 *  where eventType identifies the type of the JavaScript event generated:
 *    0: KeyPress
 *    1: KeyDown
 *    2: KeyUp
 *  and the other parameters come from the event object as-is.
 */
extern void emscripten_set_keypress_callback(void (*func)(int eventType, int charCode, int keyCode, int which, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, bool repeat));
extern void emscripten_set_keydown_callback(void (*func)(int eventType, int charCode, int keyCode, int which, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, bool repeat));
extern void emscripten_set_keyup_callback(void (*func)(int eventType, int charCode, int keyCode, int which, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, bool repeat));

/*
 * Registers a callback function for receiving browser-generated mouse input events.
 * See https://developer.mozilla.org/en/DOM/MouseEvent
 *
 * The contents of the event objects are passed to a C callback function of form
 *    void OnMouseEvent(int eventType, long screenX, long screenY, long clientX, long clientY, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, unsigned short button, unsigned short buttons);
 *  where eventType identifies the type of the JavaScript event generated:
 *    3: Click
 *    4: MouseDown
 *    5: MouseUp
 *    6: DblClick
 *  and the other parameters come from the event object as-is.
 */
extern void emscripten_set_click_callback(void (*func)(int eventType, long screenX, long screenY, long clientX, long clientY, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, unsigned short button, unsigned short buttons));
extern void emscripten_set_mousedown_callback(void (*func)(int eventType, long screenX, long screenY, long clientX, long clientY, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, unsigned short button, unsigned short buttons));
extern void emscripten_set_mouseup_callback(void (*func)(int eventType, long screenX, long screenY, long clientX, long clientY, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, unsigned short button, unsigned short buttons));
extern void emscripten_set_dblclick_callback(void (*func)(int eventType, long screenX, long screenY, long clientX, long clientY, bool ctrlKey, bool shiftKey, bool altKey, bool metaKey, unsigned short button, unsigned short buttons));

/*
 * This macro-looking function will cause Emscripten to
 * generate a comment in the generated code.
 * XXX This is deprecated for now, because it requires us to
 *     hold all global vars in memory. We need a better solution.
 */
//extern void EMSCRIPTEN_COMMENT(const char *text);

/*
 * Profiling tools.
 * INIT must be called first, with the maximum identifier that
 * will be used. BEGIN will add some code that marks
 * the beginning of a section of code whose run time you
 * want to measure. END will finish such a section. Note: If you
 * call begin but not end, you will get invalid data!
 * The profiling data will be written out if you call Profile.dump().
 */
extern void EMSCRIPTEN_PROFILE_INIT(int max);
extern void EMSCRIPTEN_PROFILE_BEGIN(int id);
extern void EMSCRIPTEN_PROFILE_END(int id);

#ifdef __cplusplus
}
#endif

