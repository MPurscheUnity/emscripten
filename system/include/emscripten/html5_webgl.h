/*
 * Copyright 2020 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

#include <webgl/webgl1.h>

#ifdef __cplusplus
#if !defined(__DEFINED_pthread_t)
typedef unsigned long pthread_t;
#define __DEFINED_pthread_t
#endif
#else
#if !defined(__DEFINED_pthread_t)
typedef struct __pthread * pthread_t;
#define __DEFINED_pthread_t
#endif
#endif

#ifdef __cplusplus
extern "C" {
#endif

#ifndef EM_BOOL
#define EM_BOOL int
#endif

#ifndef EMSCRIPTEN_RESULT
#define EMSCRIPTEN_RESULT int
#endif

#define EMSCRIPTEN_EVENT_WEBGLCONTEXTLOST      31
#define EMSCRIPTEN_EVENT_WEBGLCONTEXTRESTORED  32

typedef int EMSCRIPTEN_WEBGL_CONTEXT_HANDLE;

typedef int EMSCRIPTEN_WEBGL_CONTEXT_PROXY_MODE;
#define EMSCRIPTEN_WEBGL_CONTEXT_PROXY_DISALLOW 0
#define EMSCRIPTEN_WEBGL_CONTEXT_PROXY_FALLBACK 1
#define EMSCRIPTEN_WEBGL_CONTEXT_PROXY_ALWAYS   2

typedef int EM_WEBGL_POWER_PREFERENCE;
#define EM_WEBGL_POWER_PREFERENCE_DEFAULT 0
#define EM_WEBGL_POWER_PREFERENCE_LOW_POWER 1
#define EM_WEBGL_POWER_PREFERENCE_HIGH_PERFORMANCE 2

typedef struct EmscriptenWebGLContextAttributes {
  EM_BOOL alpha;
  EM_BOOL depth;
  EM_BOOL stencil;
  EM_BOOL antialias;
  EM_BOOL premultipliedAlpha;
  EM_BOOL preserveDrawingBuffer;
  EM_WEBGL_POWER_PREFERENCE powerPreference;
  EM_BOOL failIfMajorPerformanceCaveat;

  int majorVersion;
  int minorVersion;

  EM_BOOL enableExtensionsByDefault;
  EM_BOOL explicitSwapControl;
  EMSCRIPTEN_WEBGL_CONTEXT_PROXY_MODE proxyContextToMainThread;
  EM_BOOL renderViaOffscreenBackBuffer;
} EmscriptenWebGLContextAttributes;

extern void emscripten_webgl_init_context_attributes(EmscriptenWebGLContextAttributes *attributes);

extern EMSCRIPTEN_WEBGL_CONTEXT_HANDLE emscripten_webgl_create_context(const char *target, const EmscriptenWebGLContextAttributes *attributes);

extern EMSCRIPTEN_RESULT emscripten_webgl_make_context_current(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);

extern EMSCRIPTEN_WEBGL_CONTEXT_HANDLE emscripten_webgl_get_current_context(void);

extern EMSCRIPTEN_RESULT emscripten_webgl_get_drawing_buffer_size(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context, int *width, int *height);

extern EMSCRIPTEN_RESULT emscripten_webgl_get_context_attributes(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context, EmscriptenWebGLContextAttributes *outAttributes);

extern EMSCRIPTEN_RESULT emscripten_webgl_destroy_context(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);

extern EM_BOOL emscripten_webgl_enable_extension(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context, const char *extension);

extern EM_BOOL emscripten_webgl_enable_ANGLE_instanced_arrays(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);

extern EM_BOOL emscripten_webgl_enable_OES_vertex_array_object(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);

extern EM_BOOL emscripten_webgl_enable_WEBGL_draw_buffers(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);

extern EM_BOOL emscripten_webgl_enable_WEBGL_draw_instanced_base_vertex_base_instance(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);

typedef EM_BOOL (*em_webgl_context_callback)(int eventType, const void *reserved, void *userData);
extern EMSCRIPTEN_RESULT emscripten_set_webglcontextlost_callback_on_thread(const char *target, void *userData, EM_BOOL useCapture, em_webgl_context_callback callback, pthread_t targetThread);
extern EMSCRIPTEN_RESULT emscripten_set_webglcontextrestored_callback_on_thread(const char *target, void *userData, EM_BOOL useCapture, em_webgl_context_callback callback, pthread_t targetThread);

extern EM_BOOL emscripten_is_webgl_context_lost(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);

extern EMSCRIPTEN_RESULT emscripten_webgl_commit_frame(void);

extern EM_BOOL emscripten_supports_offscreencanvas(void);

// Returns function pointers to WebGL 1 functions. Please avoid using this function ever - all WebGL1/GLES2 functions, even those for WebGL1 extensions, are available to user code via static linking. Calling GL functions
// via function pointers obtained here is slow, and using this function can greatly increase resulting compiled program size. This functionality is available only for easier program code porting purposes, but be aware
// that calling this is causing a noticeable performance and compiled code size hit.
extern void *emscripten_webgl1_get_proc_address(const char *name);

// Returns function pointers to WebGL 2 functions. Please avoid using this function ever - all WebGL2/GLES3 functions, even those for WebGL2 extensions, are available to user code via static linking. Calling GL functions
// via function pointers obtained here is slow, and using this function can greatly increase resulting compiled program size. This functionality is available only for easier program code porting purposes, but be aware
// that calling this is causing a noticeable performance and compiled code size hit.
extern void *emscripten_webgl2_get_proc_address(const char *name);

// Combines emscripten_webgl1_get_proc_address() and emscripten_webgl2_get_proc_address() to return function pointers to both WebGL1 and WebGL2 functions. Same drawbacks apply.
extern void *emscripten_webgl_get_proc_address(const char *name);

#define emscripten_set_webglcontextlost_callback(target, userData, useCapture, callback)      emscripten_set_webglcontextlost_callback_on_thread(     (target), (userData), (useCapture), (callback), EM_CALLBACK_THREAD_CONTEXT_CALLING_THREAD)
#define emscripten_set_webglcontextrestored_callback(target, userData, useCapture, callback)  emscripten_set_webglcontextrestored_callback_on_thread( (target), (userData), (useCapture), (callback), EM_CALLBACK_THREAD_CONTEXT_CALLING_THREAD)

#define GLint int
#define GLenum int

#define EMSCRIPTEN_WEBGL_PARAM_TYPE int
#define EMSCRIPTEN_WEBGL_PARAM_TYPE_INT   0
#define EMSCRIPTEN_WEBGL_PARAM_TYPE_FLOAT 1

char  *emscripten_webgl_get_supported_extensions(void);

double emscripten_webgl_get_shader_parameter_d(GLint shader, GLenum param);
char  *emscripten_webgl_get_shader_info_log_utf8(GLint shader);
char  *emscripten_webgl_get_shader_source_utf8(GLint shader);

double emscripten_webgl_get_program_parameter_d(GLint program, GLenum param);
char  *emscripten_webgl_get_program_info_log_utf8(GLint program);

double emscripten_webgl_get_vertex_attrib_d(int index, GLenum param);
GLint emscripten_webgl_get_vertex_attrib_o(int index, GLenum param);
int emscripten_webgl_get_vertex_attrib_v(int index, GLenum param, void *dst, int dstLength, EMSCRIPTEN_WEBGL_PARAM_TYPE dstType);

double emscripten_webgl_get_uniform_d(GLint program, int location);
int emscripten_webgl_get_uniform_v(GLint program, int location, void *dst, int dstLength, EMSCRIPTEN_WEBGL_PARAM_TYPE dstType);

int emscripten_webgl_get_parameter_v(GLenum param, void *dst, int dstLength, EMSCRIPTEN_WEBGL_PARAM_TYPE dstType);
double emscripten_webgl_get_parameter_d(GLenum param);
GLint emscripten_webgl_get_parameter_o(GLenum param);
char *emscripten_webgl_get_parameter_utf8(GLenum param);
void emscripten_webgl_get_parameter_i64v(GLenum param, int64_t *dst);

#undef GLint
#undef GLenum

#ifdef __cplusplus
} // ~extern "C"
#endif
