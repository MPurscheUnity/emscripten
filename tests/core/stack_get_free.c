#include <stdlib.h>
#include <stdio.h>
#include <alloca.h>
#include <emscripten.h>
#include <emscripten/stack.h>
#include <assert.h>
#include <string.h>

void __attribute__((noinline)) DoSomething(void *addr)
{
	memcpy(addr, addr + 42, 13);
}

void TestStackValidity()
{
	unsigned int used, free, free2, total;

	used = abs((int)emscripten_stack_get_base() - (int)emscripten_stack_get_current());
	free = abs((int)emscripten_stack_get_end() - (int)emscripten_stack_get_current());
	free2 = emscripten_stack_get_free();
	total = abs((int)emscripten_stack_get_end() - (int)emscripten_stack_get_base());
	assert(used + free == total);
	assert(free == free2);
}

int main()
{
	printf("Stack free: %u\n", emscripten_stack_get_free());
	TestStackValidity();

	unsigned int prevFree = emscripten_stack_get_free();
	for(int i = 0; i < 10; ++i)
	{
		void *p = alloca(emscripten_random() >= 0 ? 256*1024 : 255 * 1024);
		unsigned int free = emscripten_stack_get_free();
		assert(prevFree - free == 256*1024);
		prevFree = free;
		DoSomething(p);
		printf("Stack free: %u\n", emscripten_stack_get_free());
		TestStackValidity();
	}
#ifdef REPORT_RESULT
	REPORT_RESULT(0);
#endif
}
