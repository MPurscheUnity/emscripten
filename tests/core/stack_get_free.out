Stack free: 4194160
Stack free: 3932016
Stack free: 3669872
Stack free: 3407728
Stack free: 3145584
Stack free: 2883440
Stack free: 2621296
Stack free: 2359152
Stack free: 2097008
Stack free: 1834864
Stack free: 1572720